### PR TITLE
Utility Types for `array`, `string`, `number`, `ts`, and `collection`

### PR DESCRIPTION
**Pull request by pskfyi**
Extracted from #97.

I recommend reading this one commit at a time.

It's all typedefs - no functions, so no unit tests.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ type Indices = Tuple.Indices<["a", "b", "c"]>; // [0, 1, 2]
 type Index = Tuple.Index<["a", "b", "c"]>; // 0 | 1 | 2
 type Reversed = Tuple.Reverse<[1, 2, 3]>; // [3, 2, 1]
 type Deed = Tuple.FromIndices<["d", "e"], [0, 1, 1, 0]>; // ["d", "e", "e", "d"]
+type ThreeUnknowns = Tuple.OfLength<3>; // [unknown, unknown, unknown]
```
Comment (AustinArey) [inline readme.md:53]: 👍

File: number/types.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * type Zero = Integer<0> // 0
+ * type NegZero = Integer<-0> // 0, TS cannot distinguish between 0 and -0
+ * type Int = Integer<1> // 1
+ * type NegInt = Integer<-1> // -1 */
+export type Integer<T extends number> = NumberType<T> extends
+  `${infer _}integer` | "zero" ? T
+  : never;
+
+export declare namespace Num {
+  export { Finite, Float, Integer, NumberType as Type, Wide };
```
Comment (AustinArey) [inline number/types.ts:83]: Nice, this reads great on the import implementation!

File: number/types.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * type Zero = NumberType<0> // "zero"
+ * type NegZero = NumberType<-0> // "zero"
+ * type Inf = NumberType<1e999> // "+infinity"
+ * type NegInf = NumberType<-1e999> // "-infinity"
+ * type Flt = NumberType<1.1> // "+float"
+ * type NegFlt = NumberType<-1.1> // "-float"
+ * type Int = NumberType<1> // "+integer"
+ * type NegInt = NumberType<-1> // "-integer" */
+export type NumberType<T extends number> = T extends infer N //
+  ? number extends N // If `number` is `N`, then `T` is exactly `number`.
```
Comment (AustinArey) [inline number/types.ts:24]: I read this as "If number is a subtype of N"

File: number/types.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    ? "number"
+  : N extends number // Otherwise, `T` is a specific number.
```
Comment (AustinArey) [inline number/types.ts:26]: I read this as "If N is a subtype of number"

File: readme.md [hunk 5]
```diff
@@ -237,6 +238,31 @@ deno run --allow-read --allow-run \
...
+## `number`
+
+Number-related utilities.
+
+```ts
+import { Num } from "https://deno.land/x/handy/number/types.ts";
+
+type T = Num.Type<1.1>; // "+float"
+type U = Num.Type<0>; // "zero"
+type V = Num.Type<-5>; // "-int"
```
[the remark this answers is not in the document]
Comment (pskfyi) [reply readme.md]: Great catch! Fixed.

**Review by AustinArey — APPROVED** [review]
Looks good

**Pull request merged**